bite-size-notes/week-8-c-LQG-stability-smvial.ipynb

```diff
@@ -0,0 +1,363 @@
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "##### Build LQG\n",
+    "To construct an LQG system, we will need a $Q$ and $R$ for our cost function. We will use simple stand-ins.\n",
+    "\n",
+    "$$\n",
+    "\n",
+    "\\begin{align*}\n",
+    "R &= 1 \\\\\n",
+    "Q &= q \\begin{bmatrix}\n",
+    "1 & 1 \\\\\n",
+    "1 & 1\n",
+    "\\end{bmatrix}\n",
+    "\\end{align*}\n",
+    "\n",
+    "$$\n",
+    "\n",
+    "Here $q>0$. Great.  Now, we may begin building our LQG. We will Start by finding the cost matrix $P$ using the continuous time infinite horizon Algebraic Riccati Equation (ARE):\n",
+    "\n",
+    "$$\n",
+    "\n",
+    "\\dot{P} = A^TP + PA - PBR^{-1}B^TP + Q = 0\n",
+    "\n",
+    "$$\n",
+    "\n",
+    "It is presumed that $\\dot{P}$ goes to 0 over time. The code cell below solves the ARE for $P$ numerically.  We may then use $P$ to solve for our control gain, $K = R^{-1}B^TP$, such that $u = -Kx$. The code below also performs this step. \n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 27,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Steady state P:\n",
+      "[[9.  4.5]\n",
+      " [4.5 4.5]]\n",
+      "Control Gain K:  [-4.5 -4.5]\n",
+      "Analytically derived K:  [-4.5 -4.5]\n"
+     ]
+    }
+   ],
+   "source": [
+    "import numpy as np\n",
+    "import scipy as sp\n",
+    "\n",
+    "# Define matrices and variables\n",
+    "A = np.array([[1.0, 1.0], [0.0, 1.0]])\n",
+    "B = np.array([[0.0], [1.0]])\n",
+    "q = 2.25\n",
+    "Q = q*np.ones((2,2))\n",
+    "R = np.array([[1.0]])\n",
+    "\n",
+    "\n",
+    "# Solve the continuous time ARE for P\n",
+    "P = sp.linalg.solve_continuous_are(A, B, Q, R)\n",
+    "print(\"Steady state P:\")\n",
+    "print(P)\n",
+    "\n",
+    "# Solve for control gain K\n",
+    "K = B.T@P\n",
+    "print(\"Control Gain K: \", K[0])\n",
+    "\n",
+    "# Check against analytic solution\n",
+    "K_a = (2 + np.sqrt(4+q)) * np.ones(2)\n",
+    "print(\"Analytically derived K: \", K_a)\n",
+    "\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Boyle somehow derived an analytic solution for the control gain (presumably using the time he saved for writing an abstract).  He came up with:\n",
+    "$$\n",
+    "K = (2 + \\sqrt{4+q})\\begin{bmatrix}\n",
+    "1 & 1 \n",
+    "\\end{bmatrix}\n",
+    "$$\n",
+    "\n",
+    "The above code computes and displays this analytic solution for comparison to our numerical soluton (given same q).  We can confirm that we are on the right track. \n",
+    "\n",
+    "Next, we will do basically the same thing, but on the estimator side of our LQR.  We will solve an ARE for the steady state $\\Sigma$, and then solve for our Kallman filter gain, $L = \\Sigma C^T S_v^{-1}$. The ARE in this case is\n",
```

Comment:
I don't think sigma or S were defined though they presumably are the same values of Q and R.

Reply:
These are defined in the "defining the system" section.  However, you are correct, $S_w$ and $S_v$ correspond to what we have been calling Q and R in class.  I just disliked the overloading of Q and R as both cost function weights and random variable covariances. $\sigma_w$ and $sigma_v$ are the standard deviation of individual elements of the W and V random variable vectors.

Reply:
Aaah.  I see what you may have meant.  I introduce these variables, but I do not explicitly define them (except in code). Noted.